kernel/mmanutils.c

```diff
@@ -184,9 +267,134 @@ void* myst_mremap(
+            uint8_t* plo = p->addr;
+            uint8_t* phi = (uint8_t*)p->addr + p->length;
+            uint8_t* maxlo = _max_ptr(lo, plo);
+            uint8_t* minhi = _min_ptr(hi, phi);
+
+            /* if there is an overlap */
+            if (maxlo < minhi)
+            {
+                size_t llength = maxlo - plo;
+                size_t rlength = phi - minhi;
+                size_t roffset = p->offset + (minhi - plo);
+
+                //     .........
+                // ..........
+                //
+                // .........
+                //     ..........
+
+                // left range:  [plo:llength]
+                // right range: [maxhi:rlength]
+
+                if (llength && rlength)
+                {
+                    // printf("case1: split\n");
+                    msync_mapping_t* rm;
+
+                    /* create the right mapping */
+                    if (!(rm = _new_msync_mapping(
+                              p->fd,
+                              roffset, /* offset */
+                              minhi,   /* addr */
+                              rlength))) /* length */
+                    {
+                        ERAISE(-ENOMEM);
+                    }
+
+                    /* insert right mapping into list */
+                    rm->next = p->next;
+                    p->next = rm;
+
+                    /* update the left mapping length */
+                    p->length = llength;
+                    prev = p;
+                }
+                else if (llength)
+                {
+                    // printf("case2: left\n");
+
+                    /* update the left mapping length */
+                    p->length = llength;
+                    prev = p;
+                }
+                else if (rlength)
+                {
+                    // printf("case3: right\n");
+                    p->offset = roffset;
+                    p->addr = minhi;
+                    p->length = rlength;
+                    prev = p;
+                }
+                else
+                {
+                    // printf("case4: remove\n");
+                    if (prev)
+                        prev->next = p->next;
+                    else
+                        _msync_mappings = p->next;
+
+                    free(p);
+                }
+            }
+            else
+            {
+                prev = p;
+            }
+
+            p = next;
+        }
+    }
+    myst_spin_unlock(&_msync_mappings_lock);
+
+    goto done;
+
+done:
+    return ret;
+}
+
 int myst_munmap(void* addr, size_t length)
 {
-    int ret = myst_mman_munmap(&_mman, addr, length);
+    int ret = 0;
+
+    /* address cannot be null and must be aligned on a page boundary */
+    if (!addr || ((uint64_t)addr % PAGE_SIZE))
+        ERAISE(-EINVAL);
+
+    /* length cannot be 0 and must be aligned on a page boundary */
+    if (!length || (length % PAGE_SIZE))
```

Comment:
I think we relaxed this. The expected behavior is to align LENGTH to a multiple of the page size. See the implementation of _munmap.

Reply:
You are right, although the man page was unclear:

```
EINVAL We don't like addr, length, or offset (e.g., they are too large, or not aligned on a page boundary).
```

I wrote this program to verify:

```
#include <sys/mman.h>
#include <stdio.h>

int main()
{
    int prot = PROT_READ | PROT_WRITE;
    int flags = MAP_PRIVATE | MAP_ANONYMOUS;
    void* addr = mmap(NULL, 4097, prot, flags, -1, 0);

    printf("addr=%p\n", addr);

    return 0;
}
```

Reply:
I remember dotnet runtime does free a memory region with a length that is not aligned to page size.